Use PyTest in SQLLogic test runner

Flogex opened the pull request.

Tishj commented:
This is already working great!

Just one thing is weird, the query object is printed but it doesn't have a `__repr__` so there's a bunch of information missing:
```
=========================================================================================== FAILURES ===========================================================================================
__________________________________________________________________ test_sqllogic[test/sql/explain/test_explain_analyze.test] ___________________________________________________________________
/Users/thijs/DuckDBLabs/duckdb/test/sql/explain/test_explain_analyze.test:77 <sqllogictest.statement.query.Query object at 0x10fb8d450>

During handling of the above exception, another exception occurred:
/Users/thijs/DuckDBLabs/duckdb/test/sql/explain/test_explain_analyze.test:77 <sqllogictest.statement.query.Query object at 0x10fb8d450>
```

I'm almost 100% sure this failure is not related to this PR btw, just the python sqllogictester lagging behind the c++ implementation, or some other limitation related to the tester itself.

Kinda offtopic:
I've been meaning to change the way the tester runs, currently it causes a lot of friction because it automatically discovers new tests and tries to run them, which, at the rate that tests are added, often means at least one test every month is breaking.
Because of this I wanted to move to an accepted list of tests, which we'll manually add tests to that we've made sure are working.
Perhaps the tester can figure out the discrepancy between available tests and the accepted list to give a warning when there are a bunch of tests that have piled up that should be checked and either added to the skip list or get accepted.

Tishj commented:
Ah actually, the problem I mentioned already exists, not related to this PR:
(ran with the old script)
this:
```
test/sql/explain/test_explain_analyze.test:77 <sqllogictest.statement.query.Query object at 0x105091c10>
```

```
[0/1] test/sql/explain/test_explain_analyze.test
┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││    Query Profiling Information    ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘
select name, value from duckdb_settings() where value != 'NULL' and value != ''
┌────────────────────────────────────────────────┐
│┌──────────────────────────────────────────────┐│
││              Total Time: 0.0003s             ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
┌───────────────────────────┐
│           QUERY           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│    ────────────────────   │
│            name           │
│           value           │
│                           │
│          92 Rows          │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│           FILTER          │
│    ────────────────────   │
│  ((value != 'NULL') AND   │
│       (value != ''))      │
│                           │
│          92 Rows          │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│    ────────────────────   │
│             #1            │
│             #0            │
│                           │
│          114 Rows         │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         TABLE_SCAN        │
│    ────────────────────   │
│         Function:         │
│      DUCKDB_SETTINGS      │
│                           │
│          114 Rows         │
│          (0.00s)          │
└───────────────────────────┘



┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││    Query Profiling Information    ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘

┌────────────────────────────────────────────────┐
│┌──────────────────────────────────────────────┐│
││              Total Time: 0.0001s             ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
┌───────────────────────────┐
│           QUERY           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│    ────────────────────   │
│            true           │
│                           │
│           1 Rows          │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         DUMMY_SCAN        │
│    ────────────────────   │
│           1 Rows          │
│          (0.00s)          │
└───────────────────────────┘



┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││    Query Profiling Information    ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘
SELECT current_setting('enable_external_access')
┌────────────────────────────────────────────────┐
│┌──────────────────────────────────────────────┐│
││              Total Time: 0.0002s             ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
┌───────────────────────────┐
│           QUERY           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│    ────────────────────   │
│            true           │
│                           │
│           1 Rows          │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         DUMMY_SCAN        │
│    ────────────────────   │
│           1 Rows          │
│          (0.00s)          │
└───────────────────────────┘



test/sql/explain/test_explain_analyze.test:77 <sqllogictest.statement.query.Query object at 0x105091c10>
ERROR
```

Tishj commented on tools/pythonpkg/scripts/conftest.py:
```diff
@@ -0,0 +1,108 @@
+    pytestmark = pytest.mark.skip(reason="Keyboard interrupt")
+
+def pytest_configure(config: pytest.Config):
+    # TODO: Change working directory to temp file. Store original working directory.
```
I guess you want to do this to enable relative paths in test files?
I don't think the original python sqllogictester does this, so I wouldn't make this a blocker for this PR

Flogex committed:
Use pytest in SQLLogic Python test runner

Integrate with pytest to be able to use a fully-fledged test runner.
Adds automatic discovery of .test files.
Adds a few command line options inspired by Catch2.

Flogex committed:
Exclude slow tests from default run, don't print progress if stdout is captured, & formatting

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Add --build-dir parameter and use pytest tmp_path as test directory

Flogex committed:
Move files to tools/pythonpkg/sqllogic

Flogex committed:
Formatting & remove tests from SKIPPED_TESTS again

Flogex committed:
Use new test runner in nightly tests

Flogex committed:
Fix directory path in nightly workflow

Flogex committed:
Don't set CI environment variable explicitly

GitHub Actions runners set this automatically

Flogex committed:
Make it possible to run all tests (including slow ones) with -m 'all'.

Flogex committed:
Set import-mode in pytest.ini

With import mode prepend, the tools/pythonpkg directory is added to the sys.path.
This leads to import errors when importing the duckdb package because there is
also a directory named tools/pythonpkg/duckdb.
Hence, setting the import mode to 'importlib' ('append' would also work).

Flogex committed:
Remove sqllogictest_python.py file

Flogex committed:
Add 'test' directory as default for --test-dir

Flogex committed:
Tishj's changes

Flogex committed:
Fix default for --test-dir

When using nargs=* and action=extend and a default value, then the default value
is always prepended to the list (of directory paths, in this case)

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Formatting

Flogex committed:
Remove outdated comment

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Skip some copy tests in Python

Flogex committed:
Format skipped_tests.py

Flogex committed:
Merge commit 'dc4bb03fd1ccde4575e2fb3091787af42afe9741' into sqllogic-pytes

Last nightlyt

hawkfish committed:
Issue #16839: Disable TIMESTAMP Casts

* Add a setting to disable timestamp => timestamptz casts in ICU (for safety)

fixes: duckdb/duckdb#16839

hawkfish committed:
Issue #16839: Disable TIMESTAMP Casts

* Add test.

fixes: duckdb/duckdb#16839

hawkfish committed:
Merge branch 'main' into timestamptz-casts

hawkfish committed:
Issue #16839: Disable TIMESTAMP Casts

* PR Feedback: Better message.

fixes: duckdb/duckdb#16839

hawkfish committed:
Issue #16839: Disable TIMESTAMP Casts

* PR Feedback: Better message test...

fixes: duckdb/duckdb#16839

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

hawkfish committed:
Issue #17049: ICU Date Cast

* Add missing DATE => TIMESTAMPTZ cast

fixes: duckdb/duckdb#17049
fixes: duckdblabs/duckdb-internal#4623

Commit:
Make block header size adaptable for encryption

staticlibs committed:
Notify JDBC repo to run Vendor.yml workflow

This change adds a dispatch call to `Vendor.yml` JDBC workflow the same
way it is now done for ODBC repo. On JDBC side the workflow changes are
being added in PR duckdb/duckdb-java#172 .

carlopi committed:
Python.yml: Add back logic to perform fast-fail on Python 3.10

Minor adaptation to pass manylinux_2_28 correctly

hawkfish committed:
Internal #4667: 2025b TimeZone Data

* Add test and update script.
* Still need to deal with "Error downloading object: meta/icumeta.jar"

hawkfish committed:
Merge branch 'main' into timestamptz-casts

hawkfish committed:
Issue #17046: AsOf Left Predicates

* Handle join predicates that only reference the left side.

fixes: duckdb/duckdb#17046
fixes: duckdblabs/duckdb-internal#4621

hawkfish committed:
Merge branch 'main' into asof-pushdown

hawkfish committed:
Issue #17049: ICU Date Cast

* Add extreme date tests.

fixes: duckdb/duckdb#17049
fixes: duckdblabs/duckdb-internal#4623

hawkfish committed:
Issue #17046: AsOf Left Predicates

* Use code generator for serialisation.

fixes: duckdb/duckdb#17046
fixes: duckdblabs/duckdb-internal#4621

hawkfish committed:
Issue #17115: TimeTZ Approximate Quantile

* Fix implementation to actually work...
* Add tests.

fixes: duckdb/duckdb#17115
fixes: duckdblabs/duckdb-internal#4669

hawkfish committed:
Merge branch 'main' into icu-2025b

hawkfish committed:
Internal #4667: 2025b TimeZone Data

* Hack around clone error
* Add new TZ to test table

fixes: duckdblabs/duckdb-internal#4667

szarnyasg committed:
README: Fix to building link

hawkfish committed:
Issue #16839: Disable TIMESTAMP Casts

* Nail down ICU settings.

fixes: duckdb/duckdb#16839

Mytherin committed:
README: Fix to building link (#17161)

c-herrewijn committed:
removed duplicate code

ZuleykhaPavlichenkova-TomTom committed:
add missing pipe to the beginning of prepare-status step's run: instruction to fix the "unexpected end of command" issue

Mytherin committed:
[InvokeCI] Add missing pipe to run instruction (#17163)

The latest InvokeCI run had failed because of missing pipe in the
`prepare-status` step's `run:` instruction.
This PR add a missing pipe and it should fix the ["unexpected end of
command"
issue](https://github.com/duckdb/duckdb/actions/runs/14505429008/job/40705456895#step:2:1).

marcoslot committed:
Add filename to GZIP stream error

Mytherin committed:
Internal #4667: 2025b TimeZone Data (#17160)

* Add test and update script.
* Hack around clone error
* Add new TZ to test table

taniabogatsch committed:
eager constraint checking in update

c-herrewijn committed:
unify register function list

hawkfish committed:
Merge branch 'main' into summarize-timetz

hawkfish committed:
Issue #17115: TimeTZ Approximate Quantile

* Merge with AVG test.

fixes: duckdb/duckdb#17115
fixes: duckdblabs/duckdb-internal#4669

hawkfish committed:
Merge branch 'main' into date-timestamptz

Commit:
Pass block manager where possible

Mytherin committed:
Unify function list (#17168)

Changes copied from
https://github.com/Mytherin/duckdb/tree/unifyfunctionlist.
Applied fix to `FunctionList::RegisterExtensionFunctions`

taniabogatsch committed:
Merge branch 'main' into eager-fix

taniabogatsch committed:
format fix

Commit:
fix tidy errors

Commit:
Add missing include

Commit:
Fix string inline test

Commit:
fixing some last tests

Commit:
Fix block header sz in rowdata

Commit:
set storage version + string inline fix attempt

Commit:
Fix grouping feature with interval type

Tishj committed:
generate the EXTENSION_SECRET_TYPES instead of hardcoding them

Mytherin committed:
[Dev] Generate the `EXTENSION_SECRET_TYPES` instead of hardcoding them (#17183)

Expansion of https://github.com/duckdb/duckdb/pull/11623

Instead of hardcoding the secret types we use `duckdb_secret_types` to
generate the list, this PR can also serve as a good example of how to
add new generation functionality to the script (there are many more
catalog types that are hardcoded still)

Mytherin committed:
Fix grouping feature with interval type (#17181)

Fix #17179

Mytherin committed:
Add filename to GZIP stream error (#17166)

When querying a large list of compressed files using a wildcard, and
some of those files are invalid (e.g. 0 bytes), it can be quite
difficult to figure out which one is causing the problem.

This PR proposes to add the (first) invalid filename to the error
message when possible.

Before:
```sql
select * from read_csv('s3://marco-crunchy-data/tmp/*.csv.gz');
IO Error: Input is not a GZIP stream
```

After:
```sql
select * from read_csv('s3://marco-crunchy-data/tmp/*.csv.gz');
IO Error: Input is not a GZIP stream: s3://marco-crunchy-data/tmp/faulty.csv.gz
```

Mytherin committed:
Issue #17115: TimeTZ Approximate Quantile (#17162)

* Fix implementation to actually work...
* Add tests.

fixes: duckdb/duckdb#17115
fixes: duckdblabs/duckdb-internal#4669

Mytherin committed:
Issue #17046: AsOf Left Predicates  (#17159)

* Handle join predicates that only reference the left side.

fixes: https://github.com/duckdb/duckdb/issues/17046
fixes: https://github.com/duckdblabs/duckdb-internal/issues/4621

Commit:
refactor Allocate

Commit:
fix for tidy check

Commit:
Adding more encryption key logic + tests

Commit:
uncomment zeroverifiers method + changed comment

carlopi committed:
[Python.yml] Skip spark tests, unclear why they fail

Commit:
changing include

Mytherin committed:
[Fix] Pass delete indexes when committing updates (#17176)

Fixes https://github.com/duckdblabs/duckdb-internal/issues/4682

Commit:
extra test for copying to a database

Commit:
Attempt to fix string inline test

Commit:
Adding no alternative verify

Commit:
Force storage version to v1.0.0

Commit:
add v1.2.0 storage test for encryption

Commit:
refactor resize() in filebuffer

Mytherin committed:
Python.yml: Add back logic to perform fast-fail on Python 3.10 (#17107)

This adds back a initial step to validate stuff actually compiles,
before launching N parallel jobs.

Spark compatibility checks should also been here, but have not found a
simple way to have them working, someone else can have a look.

Mytherin committed:
Notify JDBC repo to run Vendor.yml workflow (#17099)

This change adds a dispatch call to `Vendor.yml` JDBC workflow the same
way it is now done for ODBC repo. On JDBC side the workflow changes are
being added in PR duckdb/duckdb-java#172 .

ODBC call was initially added in #16730, see details in that PR.

Mytherin committed:
Add bind_operator callback to TableFunction - allowing table functions to directly emit a LogicalOperator

Mytherin committed:
Issue #17049: ICU Date Cast (#17067)

* Add missing DATE => TIMESTAMPTZ cast

fixes: duckdb/duckdb#17049
fixes: duckdblabs/duckdb-internal#4623

Mytherin committed:
Add bind_operator callback to TableFunction - allowing table functions to directly emit a LogicalOperator (#17196)

This PR adds a new callback to table functions - `bind_operator`:

```cpp
typedef unique_ptr<LogicalOperator> (*table_function_bind_operator_t)(ClientContext &context, TableFunctionBindInput &input, idx_t bind_index, vector<string> &return_names);
```

This function is similar to `bind_replace` - but instead of returning a
new plan (that must be bound again) - the function can return an already
bound plan directly using this callback.

Mytherin committed:
[ENCRYPTION] Make block header size adaptive (#17118)

This PR adds the possibility to change the block header size, which is
necessary for block-based database file encryption. Before this, the
header of a persistent block could only consist of an 8-byte checksum.
For the encryption of blocks, we need to store some metadata (typically
a 12-byte nonce and a 16-byte tag) at the start of each block _before_
the computed checksum, since the checksum will also be encrypted.

For now, we fix the the `ENCRYPTED_BLOCK_HEADER_SIZE` to 40 bytes, which
includes the 8-byte checksum. This will be enough for the default
encryption algorithm (`AES-GCM`), which requires a 12-byte nonce and a
16-byte tag. The block header size needs to be 8-byte aligned, because
the checksum is calculated in blocks of 64-bits.

Changing the block header size should _only_ happen when a reading or
writing an encrypted database file. To trigger this, I've added some
logic to add an (empty) encryption key. Just adding `ENCRYPTION_KEY` at
attach will force the block header size to change, but note that there
is no actual encryption yet. Trying to read a database file with a
different block header size will now throw an invalid checksum error. In
a follow up PR I will add proper encryption key logic, but the
encryption flow (`ATTACH "encrypted_db.duckdb" as encrypted
(ENCRYPTION_KEY "asdf")`) will be similar.

Note that the block header of the main database header will remain 8
bytes. This is because the main header will _not_ be encrypted since (i)
it always needs to be readable and (ii) the main header typically does
not store any sensitive data. This does add somewhat more complicated
logic in the `ChecksumAndRead()` and `ChecksumAndWrite()` methods, which
is mainly because the block manager is constructed before the file
buffer is created.

Edit:

I've added some new tests + logic to properly handle the checksum
errors. Now:

- An encrypted file is rejected when read by a duckdb <  v1.3.0
- A clean error is thrown if the database file is encrypted, but no
`ENCRYPTION_KEY` is specified
- A clean error is thrown if an `ENCRYPTION_KEY` is specified, but the
database exists and is NOT encrypted

Mytherin committed:
Issue #16839: Disable TIMESTAMP Casts (#16899)

* Add a setting to disable timestamp => timestamptz casts in ICU (for
safety)

fixes: duckdb/duckdb#16839

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Disable one more test in python runner
